src/target.rs

```diff
@@ -0,0 +1,418 @@
+//! Very basic parsing of `rustc` target triples.
+//!
+//! See the `target-lexicon` crate for a more principled approach to this.
+
+use std::{env, str::FromStr};
+
+use crate::{Error, ErrorKind};
+
+/// The parts of `rustc`'s target triple.
+///
+/// See <https://doc.rust-lang.org/cargo/appendix/glossary.html#target>.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub(crate) struct Target {
+    /// The full architecture, including the subarchitecture.
+    ///
+    /// This differs from `cfg!(target_arch)`, which only specifies the
+    /// overall architecture, which is too coarse for certain cases.
+    pub full_arch: String,
+    /// The overall target architecture.
+    ///
+    /// This is the same as the value of `cfg!(target_arch)`.
+    pub arch: String,
+    /// The target vendor.
+    ///
+    /// This is the same as the value of `cfg!(target_vendor)`.
+    pub vendor: String,
+    /// The operating system, or `none` on bare-metal targets.
+    ///
+    /// This is the same as the value of `cfg!(target_os)`.
+    pub os: String,
+    /// The environment on top of the operating system.
+    ///
+    /// This is the same as the value of `cfg!(target_env)`.
+    pub env: String,
+    /// The ABI on top of the operating system.
+    ///
+    /// This is the same as the value of `cfg!(target_abi)`.
+    pub abi: String,
+}
+
+impl Target {
+    pub fn from_cargo_environment_variables() -> Result<Self, Error> {
+        let getenv = |name| {
+            // No need to emit `rerun-if-env-changed` for these variables,
+            // as they are controlled by Cargo itself.
+            #[allow(clippy::disallowed_methods)]
+            env::var(name).map_err(|err| {
+                Error::new(
+                    ErrorKind::EnvVarNotFound,
+                    format!("failed reading {name}: {err}"),
+                )
+            })
+        };
+
+        let target = getenv("TARGET")?;
+        let (full_arch, _rest) = target.split_once('-').ok_or(Error::new(
+            ErrorKind::InvalidTarget,
+            format!("target `{target}` had an unknown architecture"),
+        ))?;
+
+        let arch = getenv("CARGO_CFG_TARGET_ARCH")?;
+        let vendor = getenv("CARGO_CFG_TARGET_VENDOR")?;
+        let os = getenv("CARGO_CFG_TARGET_OS")?;
+        let env = getenv("CARGO_CFG_TARGET_ENV")?;
+        // `target_abi` was stabilized in Rust 1.78, so may not always be available.
+        let abi = if let Ok(abi) = getenv("CARGO_CFG_TARGET_ABI") {
+            abi
+        } else {
+            Self::from_str(&target)?.abi
+        };
```

Comment:
Can we fallback to `TARGET` if these `CARGO_*` env is not set please?

There might be someone depending on that behavior

Reply:
I'd _argue_ that the documentation for `fn target` says that the `TARGET` parsing only applies to build scripts, reproduced below (emphasis mine).

> Configures the target this configuration will be compiling for.
>
> This option is automatically scraped from the `TARGET` environment variable **by build scripts**, so it's not required to call this function.

But I digress, you're right that `cc` is widely used and someone probably relies on it somewhere.